include/seastar/core/gate.hh

```diff
@@ -201,7 +214,9 @@ public:
         /// Construct a \ref holder by moving another \c holder.
         /// The referenced \ref gate is unaffected, and so the
         /// move-constructor must never throw.
-        holder(holder&& x) noexcept : _g(std::exchange(x._g, nullptr)) { }
+        holder(holder&& x) noexcept : _g(std::exchange(x._g, nullptr)) {
+            debug_hold_gate();
```

Comment:
The `x` should be unlinked from the list. Probably with the help of container_type::node_algorithms::replace_node?

Reply:
The boost intrusive hook has `auto_unlink` mode so x would get unlinked when destroyed.

Reply:
Yes, technically you're perfectly correct, but then `x` happens in a "state" when it's linked, but `x._g == nullptr` which is weird (and probably puts a landmine for future)

Reply:
fixed in v2 by adding a private release_gate function that extracts the gate member and if not null, calls `debug_release_gate()` that calls boost `unlink()`.

Reply:
One more presumably invalid corner case detected, but it worked before this patch, so I'll describe.
If you move-constuct or move-assign a released or (weird, but) moved-from gate then you call debug_hold_gate() with this->_g == nullptr and crash. Before this patch this would "just work"

Reply:
> One more presumably invalid corner case detected, but it worked before this patch, so I'll describe. If you move-constuct or move-assign a released or (weird, but) moved-from gate then you call debug_hold_gate() with this->_g == nullptr and crash. Before this patch this would "just work"

fixed in v3 by linked the holder to the gate only when the `_g` gate* is not null.

Reply:
The swap-node-ish move ctor would help to avoid this extra if, but since it's debug-mode that's ok